### PR TITLE
Default set active vectors for filters

### PR DESCRIPTION
**Pull request by MatthewFlamm**
### Overview

 

This PR implements a common helper to set a default for active_vectors on a mesh.  This currently implements option #2 from https://github.com/pyvista/pyvista/issues/2334#issuecomment-1065702247.  Reproduced here:

> 1. Raise if there are no active vectors ("explicit is better than implicit").
> 2. Activate vectors if they are unambiguous, and maybe warn while doing so. Otherwise raise ("in the face of ambiguity refuse the temptation to guess").
> 3. Always activate some vector-looking data and warn the user that this happened ("practicality beats purity").

### Details

The logic used here is to set a default active_vectors only if there is one vector-like array, i.e. shape `(n, 3)`.  It is still possible that the vector-like array is not truly a vector in the vtk sense (RGB array for example), but in this case the user should be aware of what is happening since it is unambiguous what the intent is.

### CODE REVIEW

**Comment by MatthewFlamm** [comment]
This PR is set to draft with one implementation for `streamlines_from_source`.  I will continue adding in other filters that use `active_vectors` and will mark as ready for review at that point.  The core functionality of this PR is ready for comment however.

**Comment by MatthewFlamm** [comment]
cc @adeak since we have discussed this in the past, in case you are able/willing to review.

File: pyvista/utilities/helpers.py [hunk 1]
```diff
@@ -1458,3 +1458,54 @@ def cubemap(path='', prefix='', ext='.jpg'):
...
+    possible_vectors = possible_vectors_point + possible_vectors_cell
+    n_possible_vectors = len(possible_vectors_point) + len(possible_vectors_cell)
+
+    if n_possible_vectors == 1:
+        if len(possible_vectors_point) == 1:
+            mesh.set_active_vectors(possible_vectors_point[0], preference='point')
+        else:
+            mesh.set_active_vectors(possible_vectors_cell[0], preference='cell')
+    elif n_possible_vectors < 1:
+        raise ValueError("No vector-like data available.")
```
Comment (MatthewFlamm) [inline pyvista/utilities/helpers.py:1506]: I'm not sure that `ValueError` is the right Error to be raised in both cases.

Comment (adeak) [reply pyvista/utilities/helpers.py:1506]: Thanks for the ping, I've got this on my radar but I won't be able to give a closer look today.

I agree that the errors seem a bit different. If we stick to `ValueError` then that can work I guess. But we could consider defining some semantically correct `ValueError` subclasses, e.g. `MissingDataError` and `AmbiguousDataError` or whatever. (There's a school of thought that exceptions shouldn't be distinguished by error messages, and instead one should make liberal use of exception subclasses.)

Comment (MatthewFlamm) [reply pyvista/utilities/helpers.py:1506]: I like this suggestion, but I need to think about it.  I also think that having a ton of error subclasses will lead to confusion too.  The art is getting the right level of generality vs. specificity.

File: tests/test_helpers.py [hunk 2]
```diff
@@ -186,3 +186,41 @@ def test_array_association():
...
+    # Raises if no vector-like data is present
+    mesh["scalar_data"] = np.ones((mesh.n_points, 1))
+    with pytest.raises(ValueError):
+        pyvista.set_default_active_vectors(mesh)
+    assert mesh.active_vectors_name is None
+    mesh.clear_data()
+
+    # Raises if multiple vector-like data is present
+    mesh["vec_data1"] = np.ones((mesh.n_points, 3))
+    mesh["vec_data2"] = np.ones((mesh.n_points, 3))
```
Comment (tkoyama010) [inline tests/test_helpers.py:223]:
```suggestion
    mesh["vec_data1"] = np.ones((mesh.n_points, 3))
    mesh["vec_data2"] = np.ones((mesh.n_cells, 3))
```
To test this line it is better to set `vec_data2` as cell data.
```python
n_possible_vectors = len(possible_vectors_point) + len(possible_vectors_cell)
```

Comment (MatthewFlamm) [reply tests/test_helpers.py:223]: Good idea.  Instead of changing the existing lines, I added another section to test this

**Comment by MatthewFlamm** [comment]
The glyph filter unfortunately currently has a default to orient by the active vectors.  If none are present, it simply does not orient and continues silently.  We can either allow an error to be raised as in the current PR state, or we can add in an optional value to the helper to change the error to a warning.  I'm leaning towards the second option so as to avoid a breaking change.

File: pyvista/utilities/helpers.py (hunk 1 above)
Comment (MatthewFlamm) [reply pyvista/utilities/helpers.py:1506]: Based on my comment below that glyph filter uses the active vector by default.  It would be very useful to have specific Errors for the different types of problems here.  Then we can catch those in `glyph` and provide an informative warning message.

File: pyvista/errors.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+"""Pyvista specific errors."""
```
Comment (MatthewFlamm) [inline pyvista/errors.py:1]: The only way I could figure out how to avoid circular imports was to create a new module here.  IMO, most of the errors in `pyvista.core.errors` are not related to `pyvista.core` and could be moved up.  I'm not sure if that would be a breaking change, however.

Comment (adeak) [reply pyvista/errors.py:1]: I guess we could try providing deprecated aliases in `pyvista.core` assuming we can get around the circular import problem.

**Comment by adeak** [comment]
> We can either allow an error to be raised as in the current PR state, or we can add in an optional value to the helper to change the error to a warning. I'm leaning towards the second option so as to avoid a breaking change.

Since glyphing doesn't always need vectors (see also the current behaviour), I would also want to keep the option where no vectors are present. One could easily have a point cloud they want to colour/scale by scalars only, for instance. But if I understood your edit correctly, this is what you ended up doing (still haven't looked at the implementation yet).

**Comment by MatthewFlamm** [comment]
> Since glyphing doesn't always need vectors (see also the current behaviour), I would also want to keep the option where no vectors are present

The problem is that the default behavior is to require vectors (`orient=True`), which would throw an error when no vectors are present, if we don't catch the error in glyph.  But since the behavior defaults to no orientation when no active vectors are set, there are probably users who are unaware of this and are not using an orientation vector with the default `glyph()` behavior.  Additionally, I think orienting glyphs is a very common use case (arrow vector plots being an example), so there are probably lots of users using the default behavior with orientation as well.  It is a bit of a pickle.

The options I see are, with my perceived pros and cons.  As you can see, I am a proponent of option 2.

1. Raise an error in glyph when no default vector can be set -> unexpected breaking change for some users
2. Catch errors in glyph and instead throw a warning -> keep usability and warn users when orient is turned off.  This is currently implemented.  It has the least side effects for users IMO.
3. Change default behavior to 'orient=False' and throw an error -> errors are only thrown when users specifically request `orient=True`, so it is clear that the users are intending to orient
4. Change default from `orient=True` to `orient=None`, and warn users that the default will change to False in v0.X.0.  Raise an error for no active vectors only when we switch to `orient=False` default.  This will result in extra maintenance for users who use the default glyph with orientation.

**Comment by MatthewFlamm** [comment]
I have implemented in all the filters I could find using vectors, at least ones with input parameters named `vectors`.

This is fully ready for review.

File: pyvista/errors.py (hunk 3 above)
Comment (MatthewFlamm) [reply pyvista/errors.py:1]: Let's leave out from this PR moving and deprecating the existing errors in `pyvista.core.errors`.  I mainly brought this up to justify making my life easier to create a new module here 😄 .

**Comment by adeak** [comment]
> Additionally, I think orienting glyphs is a very common use case (arrow vector plots being an example), so there are probably lots of users using the default behavior with orientation as well. It is a bit of a pickle.

I don't see it as that much of a pickle. We could choose to ignore `orient` altogether when there are no vectors, assuming that the user didn't really mean to orient if there's nothing to orient by.

> The options I see are, with my perceived pros and cons. As you can see, I am a proponent of option 2.
> 
>     1. Raise an error in glyph when no default vector can be set -> unexpected breaking change for some users
> 
>     2. Catch errors in glyph and instead throw a warning -> keep usability and warn users when orient is turned off.  This is currently implemented.  It has the least side effects for users IMO.
> 
>     3. Change default behavior to 'orient=False' and throw an error -> errors are only thrown when users specifically request `orient=True`, so it is clear that the users are intending to orient
> 
>     4. Change default from `orient=True` to `orient=None`, and warn users that the default will change to False in v0.X.0.  Raise an error for no active vectors only when we switch to `orient=False` default.  This will result in extra maintenance for users who use the default glyph with orientation.

As we both agree we shouldn't raise when there are no vectors. This almost always means someone not touching `orient` because they aren't even thinking about orienting anything. This rules out 1. and 2. Personally, I don't think we should change the default to `False`, because when you _do_ have vectors, you usually want to orient too.

From the above options I'd also go with 2 like you did. But I would consider another, additional change (which might be too eager for its own good): switching to `orient=None`, with the default staying `True`, and only raising when `orient=True` was explicitly passed and it can't be reconciled. This would mean that we don't break user code which is `orient=True` by default, but raise for users who explicitly want to orient but can't.

File: pyvista/core/filters/data_set.py [hunk 4]
```diff
@@ -1991,6 +1992,17 @@ def glyph(
...
+            try:
+                pyvista.set_default_active_vectors(self)
+            except MissingDataError:
+                warnings.warn("No vector-like data to use for orient. orient will be set to False.")
+                orient = False
+            except AmbiguousDataError as err:
+                warnings.warn(
+                    f"{err}\nIt is unclear which one to use. orient will be set to False."
+                )
+                orient = False
```
Comment (adeak) [inline pyvista/core/filters/data_set.py:2005]: And in this case, i.e. `orient=True` and there are too many potential vectors to use, I might even consider raising. This has to be a user error, and we have to force the user to do something (either disable orient, or choose vectors to orient by).

**Comment by MatthewFlamm** [comment]
> From the above options I'd also go with 2 like you did. But I would consider another, additional change (which might be too eager for its own good): switching to `orient=None`, with the default staying `True`, and only raising when `orient=True` was explicitly passed and it can't be reconciled.

I'm leaning towards agreeing with your parenthetical, but I don't lean too far in that direction. Users might be puzzled why using `glypyh(orient=True)` differs from `glyph()` unless we spend a lot of lines in the docstring explaining this.  It seems more straightforward to me to have a single behavior.

I will wait for more input on this before making more changes.

File: pyvista/utilities/helpers.py [hunk 5]
```diff
@@ -1458,3 +1459,56 @@ def cubemap(path='', prefix='', ext='.jpg'):
...
+
+def set_default_active_vectors(mesh: 'pyvista.DataSet') -> None:
+    """Set a default vectors on mesh, if not already set.
+
+    If an active vector already exists, no changes are made.
+
+    If an active vectors does not exist, it checks for possibly cell
+    or point arrays with shape ``(n, 3)``.  If only one exists, then
+    it is set as the active vectors.  If none or more than one exists,
+    then a ValueError is raised.
```
Comment (adeak) [inline pyvista/utilities/helpers.py:1472]: Aside from the markup, this still says "ValueError". Which is correct, because both errors are ValueErrors. Just pointing this out in case this phrasing was left over from the earlier implementation and you want to change it.
```suggestion
    then a ``ValueError`` is raised.
```

File: pyvista/utilities/helpers.py [hunk 6]
```diff
@@ -1458,3 +1459,56 @@ def cubemap(path='', prefix='', ext='.jpg'):
         texture.SetInputDataObject(i, flip.GetOutput())
 
     return texture
+
+
+def set_default_active_vectors(mesh: 'pyvista.DataSet') -> None:
+    """Set a default vectors on mesh, if not already set.
```
Comment (adeak) [inline pyvista/utilities/helpers.py:1465]: I know you mean "vectors" as an abstract entity, but the singular-plural combo reads surprisingly. I would either use "set default vectors" or "set default vectors data" or "set a default vectors array". There's a similar instance in the third paragraph but that's less critical as it isn't in the header line.

File: pyvista/utilities/helpers.py [hunk 7]
```diff
@@ -1458,3 +1459,56 @@ def cubemap(path='', prefix='', ext='.jpg'):
...
+    or point arrays with shape ``(n, 3)``.  If only one exists, then
+    it is set as the active vectors.  If none or more than one exists,
+    then a ValueError is raised.
+
+    Parameters
+    ----------
+    mesh : pyvista.DataSet
+        Dataset to set default active vectors.
+
+     Raises
```
Comment (adeak) [inline pyvista/utilities/helpers.py:1479]:
```suggestion
    Raises
```

File: pyvista/utilities/helpers.py [hunk 8]
```diff
@@ -1458,3 +1459,56 @@ def cubemap(path='', prefix='', ext='.jpg'):
...
+    ------
+    MissingDataError
+        If no vector-like arrays exist.
```
Comment (adeak) [inline pyvista/utilities/helpers.py:1482]:
```suggestion
        If no vector-like arrays exist.

```

File: pyvista/utilities/helpers.py [hunk 9]
```diff
@@ -1458,3 +1459,56 @@ def cubemap(path='', prefix='', ext='.jpg'):
...
+
+    possible_vectors_point = [
+        name for name, value in point_data.items() if value.ndim == 2 and value.shape[1] == 3
+    ]
+    possible_vectors_cell = [
+        name for name, value in cell_data.items() if value.ndim == 2 and value.shape[1] == 3
+    ]
+
+    possible_vectors = possible_vectors_point + possible_vectors_cell
+    n_possible_vectors = len(possible_vectors_point) + len(possible_vectors_cell)
```
Comment (adeak) [inline pyvista/utilities/helpers.py:1501]:
```suggestion
    n_possible_vectors = len(possible_vectors)
```

File: pyvista/utilities/helpers.py [hunk 10]
```diff
@@ -1458,3 +1459,56 @@ def cubemap(path='', prefix='', ext='.jpg'):
...
+
+    if n_possible_vectors == 1:
+        if len(possible_vectors_point) == 1:
+            mesh.set_active_vectors(possible_vectors_point[0], preference='point')
+        else:
+            mesh.set_active_vectors(possible_vectors_cell[0], preference='cell')
```
Comment (adeak) [inline pyvista/utilities/helpers.py:1507]: Might be a bit simpler
```suggestion
        if len(possible_vectors_point) == 1:
            preference = 'point'
        else:
            preference = 'cell'
        mesh.set_active_vectors(possible_vectors[0], preference=preference)
```

File: pyvista/utilities/helpers.py [hunk 11]
```diff
@@ -1458,3 +1459,56 @@ def cubemap(path='', prefix='', ext='.jpg'):
...
+    elif n_possible_vectors < 1:
+        raise MissingDataError("No vector-like data available.")
+    elif n_possible_vectors > 1:
+        raise AmbiguousDataError(
+            f"Multiple vector-like data available: {possible_vectors}.\n"
+            "Set one as active using DataSet.set_active_vectors(name)"
```
Comment (adeak) [inline pyvista/utilities/helpers.py:1513]: I wonder how confusing this would be if there's a single point vectors and a single cell vectors with the same name (might actually be more common than expected). We'd print the same name twice, and the `set_active_vectors` call would probably also need a `preference` kwarg to distinguish the two.

But I have no idea how to make this more descriptive for this case while keeping it concise.

File: tests/test_filters.py [hunk 12]
```diff
@@ -631,7 +631,15 @@ def test_glyph(datasets, sphere):
         geom=geoms, scale='arr', orient='Normals', factor=0.1, tolerance=0.1, progress_bar=True
     )
     assert sphere.glyph(geom=geoms[:1], indices=[None], progress_bar=True)
-    assert sphere_sans_arrays.glyph(geom=geoms, progress_bar=True)
+
+    with pytest.warns(Warning):  # tries to orient but no orientation vector available
```
Comment (adeak) [inline tests/test_filters.py:635]: I don't want to go overboard, but do we want to define a custom `DataWarning`/`ActiveDataWarning` or something to be emitted in these cases? :smile: (I'm still not great at naming things.)

File: pyvista/utilities/helpers.py [hunk 13]
```diff
@@ -1458,3 +1459,56 @@ def cubemap(path='', prefix='', ext='.jpg'):
         texture.SetInputDataObject(i, flip.GetOutput())
 
     return texture
+
+
+def set_default_active_vectors(mesh: 'pyvista.DataSet') -> None:
```
Comment (adeak) [inline pyvista/utilities/helpers.py:1464]: Since we have `import pyvista` at the top this can be the actual type, right?
```suggestion
def set_default_active_vectors(mesh: pyvista.DataSet) -> None:
```

File: tests/test_helpers.py [hunk 14]
```diff
@@ -186,3 +186,49 @@ def test_array_association():
...
+    mesh.clear_data()
+
+    # Cell data vectors
+    mesh["vec_cell"] = np.ones((mesh.n_cells, 3))
+    pyvista.set_default_active_vectors(mesh)
+    assert mesh.active_vectors_name == "vec_cell"
+    mesh.clear_data()
+
+    # Raises if no data is present
+    with pytest.raises(ValueError):
```
Comment (adeak) [inline tests/test_helpers.py:210]: Should we specify the error subclasses here?

File: tests/test_helpers.py [hunk 15]
```diff
@@ -186,3 +186,49 @@ def test_array_association():
...
+    mesh["vec_data1"] = np.ones((mesh.n_points, 3))
+    mesh["vec_data2"] = np.ones((mesh.n_points, 3))
+    with pytest.raises(ValueError):
+        pyvista.set_default_active_vectors(mesh)
+    assert mesh.active_vectors_name is None
+    mesh.clear_data()
+
+    # Raises if multiple vector-like data in cell and point
+    mesh["vec_data1"] = np.ones((mesh.n_points, 3))
+    mesh["vec_data2"] = np.ones((mesh.n_cells, 3))
```
Comment (adeak) [inline tests/test_helpers.py:231]: I would use the same array name for both, just for some edge case paranoia.
```suggestion
    mesh["vec_data"] = np.ones((mesh.n_points, 3))
    mesh["vec_data"] = np.ones((mesh.n_cells, 3))
```
Not that I see anything in the implementation that would fail if the names are the same.

File: pyvista/utilities/helpers.py (hunk 13 above)
Comment (MatthewFlamm) [reply pyvista/utilities/helpers.py:1464]: This is what I did first and it leads to a circular import in a way I don't understand. This might only be true for earlier python versions, e.g. 3.7.  I could try using `TYPE_CHECKING` and `from pyvista import DataSet` if this preferred.

File: pyvista/utilities/helpers.py (hunk 5 above)
Comment (MatthewFlamm) [reply pyvista/utilities/helpers.py]: We don't need to specify the error at all here, since it is laid out in detail later.

File: pyvista/utilities/helpers.py (hunk 6 above)
Comment (MatthewFlamm) [reply pyvista/utilities/helpers.py]: I took your suggestion.

File: pyvista/utilities/helpers.py (hunk 11 above)
Comment (MatthewFlamm) [reply pyvista/utilities/helpers.py:1514]: We don't need to set a preference to call this function because we have, so far, taken the decision to raise an error if there is more than one vector-like array.

It would be most complete to print the vector-like arrays separately for points and cells.  But we don't do this for usage like `mesh.array_names` so I don't know.

File: tests/test_helpers.py (hunk 14 above)
Comment (MatthewFlamm) [reply tests/test_helpers.py]: Done

File: tests/test_helpers.py (hunk 15 above)
Comment (MatthewFlamm) [reply tests/test_helpers.py:232]: I added a new section to the test instead of replacing.

File: tests/test_filters.py (hunk 12 above)
Comment (MatthewFlamm) [reply tests/test_filters.py:635]: I'm thinking that the warning should indicate that the input parameter is changed/ignored.  This would be a general and fairly useful warning type.  I'm blanking on a concise name too.

**Comment by MatthewFlamm** [comment]
I'm starting to think that we may want to leave glyph out of this PR, or widen the scope of this PR to include scalars.  `glyph` bad

File: pyvista/utilities/helpers.py (hunk 11 above)
Comment (adeak) [reply pyvista/utilities/helpers.py:1514]: With the preference I just meant the recommendation in the error message: "Set one as active using `DataSet.set_active_vectors(name)`". But again it would be cumbersome to try and add the preference distinction to the message (to the effect of "also don't forget to choose the right preference if you have a name clash").

File: pyvista/utilities/helpers.py (hunk 13 above)
Comment (adeak) [reply pyvista/utilities/helpers.py:1464]: Hmm, I guess it's that there's still a circular import, but the partially initialised module doesn't cause a problem because we aren't accessing any of the undefined names here. And using `pyvista.DataSet` eagerly violates that.

There's nothing wrong with the string annotation, I just thought we don't need it stringified. Let's leave it like this then, and maybe remove all the string annotations in the distant future when lazy evaluation is the default for all supported Python versions.

File: pyvista/utilities/helpers.py (hunk 11 above)
Comment (MatthewFlamm) [reply pyvista/utilities/helpers.py:1514]: We could print cell and point vectors separately and then use `DataSet.set_active_vectors(name, preference=type)` something like this.  This gives the user two pointers without having to be super explicit.

File: tests/test_filters.py (hunk 12 above)
Comment (MatthewFlamm) [reply tests/test_filters.py:635]: I am torn in two directions.

One is something like `ActiveAttributesWarning`, since we cannot set an active attribute in the vtk sense. If this too vtkish, I like your `ActiveDataWarning`.  These are the cause of the warning.

To go another way, something like `ParameterChangeWarning` to indicate the effect of what is happening.

**Comment by akaszynski** [comment]
@MatthewFlamm, is this ready for review?

**Comment by MatthewFlamm** [comment]
> @MatthewFlamm, is this ready for review?

Yes

File: tests/test_filters.py (hunk 12 above)
Comment (akaszynski) [reply tests/test_filters.py:635]: I think we can resolve this later, but as-is, this is sufficient.

**Review by akaszynski — APPROVED** [review]
Great work, and thanks for the review @adeak

**Review by adeak — APPROVED** [review]
Potential future work in
1. https://github.com/pyvista/pyvista/pull/2433#discussion_r844260617
2. https://github.com/pyvista/pyvista/pull/2433#discussion_r846012109
3. https://github.com/pyvista/pyvista/pull/2433#discussion_r849837946

But I agree with @akaszynski that this is good as-is. Thanks!

**Comment by akaszynski** [comment]
> But I agree with @akaszynski that this is good as-is. Thanks!

Trying to avoid PRs becoming too long lived (big believer in trunk based development and short lived branches).

Agree with follow-ups.
